File: medium/427. Construct Quad Tree/427. Construct Quad Tree.go
/**
 * Definition for a QuadTree node.
 */
package main

import "fmt"

type Node struct {
	Val         bool
	IsLeaf      bool
	TopLeft     *Node
	TopRight    *Node
	BottomLeft  *Node
	BottomRight *Node
}

func quad_tree(grid [][]int, x1, x2, y1, y2 int) *Node {
	if x1 == x2 && y1 == y2 {
		return &Node{Val: grid[y1][x1] == 1, IsLeaf: true}
	}

	xMiddle := (x1 + x2) / 2
	yMiddle := (y1 + y2) / 2

	topLeft := quad_tree(grid, x1, xMiddle, y1, yMiddle)
	topRight := quad_tree(grid, xMiddle+1, x2, y1, yMiddle)
	bottomLeft := quad_tree(grid, x1, xMiddle, yMiddle+1, y2)
	bottomRight := quad_tree(grid, xMiddle+1, x2, yMiddle+1, y2)

	if topLeft.IsLeaf && topRight.IsLeaf && bottomLeft.IsLeaf && bottomRight.IsLeaf {
		if topLeft.Val == topRight.Val && topRight.Val == bottomLeft.Val && bottomLeft.Val == bottomRight.Val {
			return &Node{Val: topLeft.Val, IsLeaf: true}
		}
	}

	return &Node{Val: true, IsLeaf: false, TopLeft: topLeft, TopRight: topRight, BottomLeft: bottomLeft, BottomRight: bottomRight}
}

func construct(grid [][]int) *Node {
	return quad_tree(grid, 0, len(grid)-1, 0, len(grid)-1)
}

func main() {
	grid := [][]int{{1, 1, 1, 1, 0, 0, 0, 0}, {1, 1, 1, 1, 0, 0, 0, 0}, {1, 1, 1, 1, 1, 1, 1, 1}, {1, 1, 1, 1, 1, 1, 1, 1}, {1, 1, 1, 1, 0, 0, 0, 0}, {1, 1, 1, 1, 0, 0, 0, 0}, {1, 1, 1, 1, 0, 0, 0, 0}, {1, 1, 1, 1, 0, 0, 0, 0}}

	fmt.Println(construct(grid))
}

File: medium/427. Construct Quad Tree/427. Construct Quad Tree.ts
/**
 * Definition for _Node.
 */

class _Node {
    val: boolean
    isLeaf: boolean
    topLeft: _Node | null
    topRight: _Node | null
    bottomLeft: _Node | null
    bottomRight: _Node | null
    constructor(val?: boolean, isLeaf?: boolean, topLeft?: _Node, topRight?: _Node, bottomLeft?: _Node, bottomRight?: _Node) {
        this.val = (val === undefined ? false : val)
        this.isLeaf = (isLeaf === undefined ? false : isLeaf)
        this.topLeft = (topLeft === undefined ? null : topLeft)
        this.topRight = (topRight === undefined ? null : topRight)
        this.bottomLeft = (bottomLeft === undefined ? null : bottomLeft)
        this.bottomRight = (bottomRight === undefined ? null : bottomRight)
    }
}


function construct(grid: number[][]): _Node | null {
    const quadTree = (
        x1: number, x2: number,
        y1: number, y2: number
    ): _Node | null => {
        if (x1 === x2 && y1 === y2) {
            return new _Node(!!grid[y1][x1], true)
        }

        const xMiddle = Math.floor((x1 + x2) / 2), yMiddle = Math.floor((y1 + y2) / 2)

        const topLeft = quadTree(x1, xMiddle, y1, yMiddle) || undefined
        const topRight = quadTree(xMiddle + 1, x2, y1, yMiddle) || undefined
        const bottomLeft = quadTree(x1, xMiddle, yMiddle + 1, y2) || undefined
        const bottomRight = quadTree(xMiddle + 1, x2, yMiddle + 1, y2) || undefined

        if (topLeft?.isLeaf && topRight?.isLeaf && bottomLeft?.isLeaf && bottomRight?.isLeaf) {
            if (
                topLeft.val === topRight.val &&
                topRight.val === bottomLeft.val &&
                bottomLeft.val === bottomRight.val
            ) {
                return new _Node(topLeft.val, true)
            }
        }

        return new _Node(true, false, topLeft, topRight, bottomLeft, bottomRight)
    }

    return quadTree(0, grid.length - 1, 0, grid.length - 1)
};